clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```
@@ -218,18 +490,17 @@ public CompletableFuture<Void> resetPositionsIfNeeded() {
      *
      * <p/>
      *
-     * When a response is received, positions are validated and, if a log truncation is
-     * detected, a {@link LogTruncationException} will be saved in memory, to be thrown on the
+     * When a response is received, positions are validated and, if a log truncation is detected, a
+     * {@link LogTruncationException} will be saved in memory in cachedUpdatePositionsException, to be thrown on the
      * next call to this function.
      */
-    public CompletableFuture<Void> validatePositionsIfNeeded() {
-        Map<TopicPartition, SubscriptionState.FetchPosition> partitionsToValidate =
-                offsetFetcherUtils.getPartitionsToValidate();
+    void validatePositionsIfNeeded() {
+        Map<TopicPartition, SubscriptionState.FetchPosition> partitionsToValidate = offsetFetcherUtils.getPartitionsToValidate();
```

`offsetFetcherUtils.getPartitionsToValidate` can throw exception from cache, so should we propagate the exception by `CompletableFuture`?

---

you're right, and that's the intention but it's done on the catch further down (exception here would short-circuit the flow as we want, and complete exceptionally on catch). 
https://github.com/apache/kafka/blob/021a29083cca99fc656b8248fab66d8e9abf05d1/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L261

Makes sense?